### ISSUE TITLE
Training Algorithm

### ISSUE THREAD

**Issue by ciccio42** [vimalabs/VIMA#12]
Hello.
First of all, thanks for sharing, and congratulations for your work. 
I want to use your model in my experiments related to object-centric Imitation Learning policy, however, I have some differences in the robot platform and simulation environment, so I need to train your model from scratch. Therefore, I have some high-level questions about your training loop. 
In particular, starting from your example.py scripts, how you use history information is clear, but how did you build the batch during the training? Particularly, how did you fill the slot related to past action-token?
I appreciate your consideration.

Francesco Rosa.

**Comment by yunfanjiang** [comment]
Hi there,

Thanks for your interest in our work. During training, we assume the access to existing dataset consisting of observation-action trajectories. The models are then trained by casually predicting action tokens. In other words, when the model is asked to predict the action at timestep t, it can only see history up to that timestep. Hope this can help!

**Comment by ciccio42** [comment]
Thanks for your reply.

To clarify with an example, let's suppose that within the batch, we have a trajectory that goes from timestamp t to timestamp t+H. First, the model predicts the action token at timestamp t, without history, because it is the first frame. Then, the model predicts the action token for timestamp t+1, with the previously predicted action token that goes in input and so on. Obviously, at each timestamp, the loss is computed, which is then averaged over the entire trajectory.   
Is it correct?

**Comment by yunfanjiang** [comment]
Yes, almost correct! The first action is predicted conditioning on first observation. Further, all actions are predicted also conditioning on prompts in our setting.

**Comment by ciccio42** [comment]
As always thanks for your reply, it is clear that you use in the history the predicted action token, but I have a two further questions:

1. During training, do you propagate in the history the predicted action token computed over the predicted action or the predicted action token computed over the target action?
2. How many past steps are included in the history?

Thanks.

**Comment by yunfanjiang** [comment]
> * During training, do you propagate in the history the predicted action token computed over the predicted action or the predicted action token computed over the target action?

Thanks for the followup questions. To answer them

1. During training, actions are predicted based on target (ground-truth) actions.
2. We feed entire history to the model during both training and test.
